src/pipeline.rs

```diff
@@ -382,3 +430,98 @@ pub(crate) fn prepare_view_targets(
         }
     }
 }
+
+#[derive(Resource, Clone)]
+pub struct PointCloudPlaybackControl {
+    pub playing: bool,
+    pub speed: f32,
+}
+impl Default for PointCloudPlaybackControl {
+    fn default() -> Self {
+        Self {
+            playing: true,
+            speed: 5.0,
+        }
+    }
+}
+
+impl ExtractResource for PointCloudPlaybackControl {
+    type Source = Self;
+
+    /// Defines how the resource is transferred into the "render world".
+    fn extract_resource(source: &Self::Source) -> Self {
+        source.clone()
+    }
+}
+
+pub fn prepare_animated_assets(
+    queue: Res<RenderQueue>,
+    mut assets: ResMut<RenderAssets<PointCloudAsset>>,
+    playback: ResMut<PointCloudPlaybackControl>,
+    time: Res<Time>,
+) {
+    if !playback.playing {
+        return;
+    }
+    for (_handle, asset) in assets.iter_mut() {
+        if let Some(animation_buffer) = asset.animation_buffer.as_mut() {
+            let mut view = vec![0.0; asset.num_points as usize * 3];
+
+            match asset.frames.as_ref().unwrap() {
+                opd_parser::Frames::I8(frames) => {
+                    let current_frame = &frames[asset.current_animation_frame];
+                    asset.animation_time += time.delta_seconds() * playback.speed;
+
+                    if current_frame.time / 1000.0 > asset.animation_time {
+                        continue;
+                    }
+
+                    asset.current_animation_frame += 1;
+                    if asset.current_animation_frame >= frames.len() {
+                        asset.current_animation_frame = 0;
+                        asset.animation_time = 0.0;
+                    }
+
+                    let scale: [f32; 3] = asset.animation_scale.into();
+                    let mut iter = frames[asset.current_animation_frame]
+                        .data
+                        .iter()
+                        .enumerate();
+                    loop {
+                        let arr = [
+                            match iter.next() {
+                                Some(a) => (a.0, *a.1),
+                                None => break,
+                            },
+                            match iter.next() {
+                                Some(a) => (a.0, *a.1),
+                                None => break,
+                            },
+                            match iter.next() {
+                                Some(a) => (a.0, *a.1),
+                                None => break,
+                            },
+                        ];
```

Comment:
I think this is just `for arr in iter.chunks(3)`.

https://doc.rust-lang.org/std/primitive.slice.html#method.chunks

This does mean you can't use `enumerate`, since `chunks` requires a slice, but it's pretty easy to just manually increment an `i: usize`.

Reply:
Actually on second thought, why not just preprocess the frames during asset creation so you actually get a buffer of `Vec3` offsets?

Reply:
The reason is that the asset stores i8 instead of f32, and I don't want to pre-convert them to f32.
Also, I was originally using `array_chunks` but it's unstable. `chunks` returns a slice which requires additional bound checks in subsequent commands. IMO this is fine for now, we can revisit after `array_chunks` stabilize.

Reply:
I still think it's best to convert each frame into a `Vec<Vec3A>` before handling it. It doesn't have to be during asset loading, you can provide a method on `Frames` to extract one frame at a time.